src/main/java/com/squareup/javapoet/CodeWriter.java

```diff
@@ -272,106 +273,158 @@ private void emitLiteral(Object o) throws IOException {
    * names visible due to inheritance.
    */
   String lookupName(ClassName className) {
-    // Different package than current? Just look for an import.
-    if (!className.packageName().equals(packageName)) {
-      if (conflictsWithLocalName(className)) {
-        return className.toString(); // A local name conflicts? Use the fully-qualified name.
+
+    // Create ClassName from current stack
+    ClassName stackContext = null;
+    for (TypeSpec typeSpec : typeSpecStack) {
+        if (typeSpec.name == null) {
+          // Skip anonymous class
+          continue;
+        }
+        if (stackContext == null) {
+            stackContext = ClassName.get(packageName, typeSpec.name);
+        } else {
+            stackContext = stackContext.nestedClass(typeSpec.name);
+        }
+    }
+
+    if (stackContext == null) {
+      // There is no stack context, e.g. annotation or interface on top-level classes
+      if (className.packageName().equals(packageName)) {
+        return Util.join(".", className.simpleNames());
+      } else {
```

Comment:
You don't need `else` after `return`, and it makes the presence of this comment not seem so strange.
```java
if (foo) {
  return "foo";
}
// There's no way to know!
```